src/rgw/rgw_file.cc

```diff
@@ -1023,33 +1023,41 @@ namespace rgw {
     return false;
   }
 
-  int RGWFileHandle::readdir(rgw_readdir_cb rcb, void *cb_arg, uint64_t *offset,
+  std::ostream& operator<<(std::ostream &os,
+			   RGWFileHandle::readdir_offset const &offset)
+  {
+    if (unlikely(!! get<uint64_t*>(&offset))) {
+      uint64_t* ioff = get<uint64_t*>(offset);
```

Comment:
confused here - the first call to `get()` passes `&offset` while the others pass `offset`

also probably worth qualifying as `boost::get` - i'm guessing that this resolves to `std::get`, so i'm surprised that it compiles

Reply:
this is a wired-in idiom of boost::variant:  the versions that take a pointer argument do not throw and return either a pointer to the stored object, or nullptr if the type coercion fails;  the versions that take an offset return either a reference to the stored object or throw boost::bad_get.  The apparent alternative is to code to typeid or an interface "which()" that returns an integer index into the defined variants--pretty ugly.  I will specify boost::get.